zap/src/main/java/org/zaproxy/zap/session/ScriptBasedSessionManagementMethodType.java

```diff
@@ -0,0 +1,668 @@
+
+    public static final String SCRIPT_TYPE_SESSION = "session";
+
+    private static final ImageIcon SCRIPT_ICON_SESSION =
+            new ImageIcon(ZAP.class.getResource("/resource/icon/16/script-session.png"));
+
+    private ExtensionScript extensionScript;
+
+    public class ScriptBasedSessionManagementMethod implements SessionManagementMethod {
+
+        private ScriptWrapper script;
+
+        private Map<String, String> paramValues;
+
+        @Override
+        public boolean isConfigured() {
+            // Always configured
+            return true;
+        }
+
+        @Override
+        public SessionManagementMethodType getType() {
+            return new ScriptBasedSessionManagementMethodType();
+        }
+
+        @Override
+        public WebSession extractWebSession(HttpMessage msg) {
+            SessionScript sessionScript = getScriptInterface(script);
+            if (sessionScript != null) {
+                return sessionScript.extractWebSession(msg);
+            }
+            return new ScriptBasedSession();
+        }
+
+        @Override
+        public WebSession createEmptyWebSession() {
+            return new ScriptBasedSession();
+        }
+
+        @Override
+        public void clearWebSessionIdentifiers(HttpMessage msg) {
+            SessionScript sessionScript = getScriptInterface(script);
+            if (sessionScript != null) {
+                sessionScript.clearWebSessionIdentifiers(msg);
+            }
+        }
+
+        @Override
+        public ApiResponse getApiResponseRepresentation() {
+            Map<String, String> values = new HashMap<>();
+            values.put("methodName", API_METHOD_NAME);
+            values.put("scriptName", script.getName());
+            values.putAll(paramValues);
+            return new SessionMethodApiResponseRepresentation<>(values);
+        }
+
+        @Override
+        public void processMessageToMatchSession(HttpMessage message, WebSession session)
+                throws UnsupportedWebSessionException {
+            SessionScript sessionScript = getScriptInterface(script);
+            if (sessionScript != null) {
+                sessionScript.processMessageToMatchSession(message, session);
+            }
+        }
+
+        @Override
+        public SessionManagementMethod clone() {
+            ScriptBasedSessionManagementMethod method = new ScriptBasedSessionManagementMethod();
+            method.script = this.script;
+            method.paramValues =
+                    this.paramValues != null ? new HashMap<String, String>(this.paramValues) : null;
+            return method;
+        }
+
+        @Override
+        public String toString() {
+            return "ScriptBasedSessionManagementMethod [script="
+                    + script
+                    + ", paramValues="
+                    + paramValues
+                    + "]";
+        }
+
+        @Override
+        public boolean equals(Object obj) {
```

Comment:
Should be changed to check `script`/`paramValues`? (Or just removed?) Same for `hashCode`.